common/tasks/multi_destination_scheduler.go

```diff
@@ -0,0 +1,241 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tasks
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+)
+
+type RateLimiter interface {
+	Wait(context.Context) error
+}
+
+// LimiterOptions are options for a single destination's limiter.
+type LimiterOptions struct {
+	// Concurrency returns the concurrency limit.
+	Concurrency func() int
+	RateLimiter RateLimiter
+}
+
+// limiter acts a concurrency and rate limiter for a single destination.
+// All access to the limiter must be locked.
+type limiter[T Task] struct {
+	sync.Mutex
+	options LimiterOptions
+	// Number of reservations help by this limiter. Restricts concurrency.
+	reservations int
+	// Any tasks that exceed the concurrency limitation are buffered in this slice.
+	buffer []T
+}
+
+func newLimiter[T Task](options LimiterOptions) *limiter[T] {
+	return &limiter[T]{
+		options: options,
+	}
+}
+
+// enqueue a task in the limiter's buffer.
+func (lim *limiter[T]) enqueue(task T) {
+	lim.buffer = append(lim.buffer, task)
+}
+
+// dequeue a task from the limiter's buffer.
+func (lim *limiter[T]) dequeue() (task T, ok bool) {
+	if len(lim.buffer) > 0 {
+		task, lim.buffer = lim.buffer[0], lim.buffer[1:]
+		return task, true
+	}
+	return task, false
+}
+
+// reserve a concurrency slot. Returns true on success.
+func (lim *limiter[T]) reserve() bool {
+	if lim.reservations < lim.options.Concurrency() {
+		lim.reservations++
+		return true
+	}
+	return false
+}
+
+// release a reservation.
+func (lim *limiter[T]) release() {
+	lim.reservations--
+}
+
+// MultiDestinationSchedulerOptions are options for creating a [MultiDestinationScheduler].
+type MultiDestinationSchedulerOptions[K comparable, T Task] struct {
+	// A function to determine the destination of a task.
+	KeyFn func(T) K
+	// Limiter options for a given destination.
+	LimiterOptionsFn func(K) LimiterOptions
+	Logger           log.Logger
+}
+
+// MultDestinationScheduler for limiting concurrency and RPS (and later applying a circuit breaker) on a per destination
+// basis. A task's destination is determined based on a provided key function.
+// Accepts and buffers tasks that exceed the destination's concurrency limit to be picked up as soon as there's more
+// availability.
+type MultiDestinationScheduler[K comparable, T Task] struct {
```

Comment:
I'm considering naming this "grouped limiter" scheduler, it conveys the idea better.

Reply:
How about "partition" or "segment" instead of "destination"? 

1) Both are more common terms for subdivisions of items in a queue.
2) They don't collide with the callback "destination", which is a different concept.
3) This could be renamed `PartitionedScheduler` or `SegmentedScheduler`, which are both more succinct.
4) I think it would allow you to rename the `limiter` to something like `partition`, `queuePartition`, or `queueSegment` which I think better encompass the enqueue/dequeue behavior they have on top of rate/concurrency limiting.